### PR TITLE
Feature/reports use local records

### PR DESCRIPTION
**Pull request by opsaaaaa**
this should have probably been multiple pull requests.
Context:
document_meta is created from data from canvas. 
Que is a background task manager that is used to generate reports. 

as an auditor, i want to be able to see stats about my organizations documents.
when i am using reports_use_document_meta
and a new report is generated 
and i visit a the report. 
then i should see stats about canvas documents
when i am not using reports_use_document_meta
and a new report is generated 
and i visit a the report. 
then i should see stats about the number of documents for each of my organizations.

as an auditor I want to be able to preview the documents for my organization and sub organizations.
when i visit a report
then i should see a list of documents 
and each document should have links to preview my documents. 

as an auditor i want to be able to download my reports.
when i visit reports
and i click a download links for one of my generated reports
then i should receive a zip file

as an auditor i want to be able to change what my files are named by.
when i visit organization settings
then i should be able to change what my documents are named by. 
and the input should allow for document.name, and document.lms_course_id
when i download a report 
then the documents name should follow my naming scheme

when aws environment variables present
then zip files should be upload to aws

when aws environment variables absent
then zip files should be handled locally.

### CODE REVIEW

File: app/controllers/organizations_controller.rb [hunk 1]
```diff
@@ -216,8 +215,40 @@ def organization_params
     end
 
     if has_role 'organization_admin'
-      org_params.permit(:name, :export_type, :slug, :period_meta_key, :enable_workflows, :inherit_workflows_from_parents, :parent_id, :lms_authentication_source, :lms_authentication_id, :lms_authentication_key, :lms_info_slug, :lms_account_id, :home_page_redirect, :skip_lms_publish, :enable_shibboleth, :idp_sso_target_url, :idp_slo_target_url, :idp_entity_id, :idp_cert, :idp_cert_fingerprint, :idp_cert_fingerprint_algorithm, :authn_context, :enable_anonymous_actions, :track_meta_info_from_document, :disable_document_view,
```
Comment (opsaaaaa) [inline app/controllers/organizations_controller.rb]: this has caused more than a few merge conflicts. i am guessing git will handle it better if its multiple lines.

File: app/helpers/report_helper.rb [hunk 2]
```diff
@@ -2,24 +2,16 @@
 require 'zip'
 
 module ReportHelper
-  def self.generate_report_as_job (org_id, account_filter, params)
-    @reports = ReportArchive.where(organization_id: org_id).all
-    @report = nil;
+  def self.generate_report_as_job (org_id, account_filter, params, report_id = nil)
+    @report = ReportArchive.find(report_id) if report_id
```
Comment (llamapope) [inline app/helpers/report_helper.rb]: check if there is already a report with the same params, if so, use that report so we don't have duplicate reports

File: app/helpers/report_helper.rb [hunk 3]
```diff
@@ -32,88 +24,101 @@ def self.generate_report (org_slug, account_filter, params, report_id)
       docs = Document.where(workflow_step_id: WorkflowStep.where(organization_id: @organization.parents.push(@organization.id), step_type: "end_step").pluck(:id), organization_id: @organization.id).where('updated_at != created_at').all
     end
     # get the report data (slow process... only should run one at a time)
-    puts 'Getting Document Meta'
-    if @organization.root_org_setting("enable_workflow_report")
-      @report_data = self.get_workflow_document_meta docs&.pluck(:id)
+    @report.used_document_meta = @organization.setting("reports_use_document_meta")
+    if @organization.setting("reports_use_document_meta")
+      puts 'Getting Document Meta'
```
Comment (opsaaaaa) [inline app/helpers/report_helper.rb]: remove puts

File: app/helpers/report_helper.rb [hunk 4]
```diff
@@ -32,88 +24,101 @@ def self.generate_report (org_slug, account_filter, params, report_id)
...
       end
       docs.each do |doc|
-        identifier = doc.id
-        folder = nil
-        folder = "#{doc.period&.slug}/" if @organization.root_org_setting("enable_workflow_report")
-        identifier = doc.name.gsub(/[^A-Za-z0-9]+/, '_') if doc.name
-        if doc.lms_course_id
-          identifier = "#{doc.lms_course_id}".gsub(/[^A-Za-z0-9]+/, '_')
-        end
+        doc_report_data = (report_data.select {|rep| rep["document_id"] == doc.id }).first
```
Comment (opsaaaaa) [inline app/helpers/report_helper.rb]: default to lms_course_id

File: app/models/organization.rb [hunk 5]
```diff
@@ -27,6 +28,16 @@ def self.export_types
     ["default","Program Outcomes"]
   end
   validates :export_type, :inclusion=> { :in => self.export_types }
+  
+  def self.name_reports_by_options
```
Comment (opsaaaaa) [inline app/models/organization.rb]: document.lms_course_id
document.remote_identifier
document.name_document

File: app/controllers/workflow_documents_controller.rb [hunk 6]
```diff
@@ -22,7 +22,7 @@ def index
     @workflow_steps = WorkflowStep.where(organization_id: organization_ids).order('name')
     @periods = Period.where(organization_id: organization_ids).order('name')
     @documents = Document.where(organization_id:org.self_and_descendants.pluck(:id))
-      .where('documents.updated_at != documents.created_at')
+      .where(Document.not_abandoned)
```
Comment (opsaaaaa) [inline app/controllers/workflow_documents_controller.rb]: change `Document.not_abandoned` to `Document.updated`

**Pull request merged**